In [2]:
from keras.datasets import mnist
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.manifold import TSNE
import warnings
warnings.filterwarnings("ignore")

(_, _), (X, y) = mnist.load_data()

X = X.reshape(len(X), -1)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=89
)

kmeans = KMeans(n_clusters=10, random_state=89)
kmeans.fit(X_train)

out1 = np.sum(kmeans.labels_ == 1)

labels_map = {}
for c in range(10):
    vals = y_train[kmeans.labels_ == c]
    labels_map[c] = np.bincount(vals).argmax()
out2 = [labels_map[i] for i in range(10)]

train_pred = np.array([labels_map[c] for c in kmeans.labels_])
out3 = round(accuracy_score(y_train, train_pred), 3)

test_clusters = kmeans.predict(X_test)
test_pred = np.array([labels_map[c] for c in test_clusters])
out4 = round(accuracy_score(y_test, test_pred), 3)

cm = confusion_matrix(y_test, test_pred)
row9 = cm[9].copy()
row9[9] = -1
out5 = np.argmax(row9)

tsne = TSNE(n_components=2, init='random', random_state=89)
X_train_tsne = tsne.fit_transform(X_train)

kmeans_tsne = KMeans(n_clusters=10, random_state=89)
kmeans_tsne.fit(X_train_tsne)

labels_map_tsne = {}
for c in range(10):
    vals = y_train[kmeans_tsne.labels_ == c]
    labels_map_tsne[c] = np.bincount(vals).argmax()

train_pred_tsne = np.array([labels_map_tsne[c] for c in kmeans_tsne.labels_])
out6 = round(accuracy_score(y_train, train_pred_tsne), 3)

(out1, out2, out3, out4, out5, out6)


c:\Users\Misha\PycharmProjects\itmo_ml_moodle\.venv\Lib\site-packages\keras\src\export\tf2onnx_lib.py:8: FutureWarning: In the future `np.object` will be defined as the corresponding NumPy scalar.
  if not hasattr(np, "object"):


(np.int64(1198),
 [np.int64(6),
  np.int64(7),
  np.int64(4),
  np.int64(8),
  np.int64(2),
  np.int64(1),
  np.int64(3),
  np.int64(0),
  np.int64(1),
  np.int64(6)],
 0.6,
 0.603,
 np.int64(4),
 0.807)

(np.int64(1198),
 [np.int64(6),
  np.int64(7),
  np.int64(4),
  np.int64(8),
  np.int64(2),
  np.int64(1),
  np.int64(3),
  np.int64(0),
  np.int64(1),
  np.int64(6)],
 0.6,
 0.603,
 np.int64(4),
 0.729) 
 запускать в гугл коллаб